### ISSUE TITLE
Type inference of generic function required a variable

### ISSUE THREAD

**Issue by minmax** [microsoft/pyright#6141]
pyright version: git main https://github.com/microsoft/pyright/commit/0d8e18cfbd318d34389fa402d4ba287a1c9cac2b Thu Oct 12 12:49:48 2023 -0700

```python
T = TypeVar("T")
E = TypeVar("E")
P = ParamSpec("P")

class Ok(Generic[T]):
    pass
class Err(Generic[T]):
    pass

Result: TypeAlias = Ok[T] | Err[E]

def wrap_function(wrapper: Callable[P, T], wrapped: Callable[..., Any]) -> Callable[P, T]:
    return wrapper

def bind(func: Callable[[T], Result[R, RE]]) -> Callable[[Result[T, E]], Result[R, RE | E]]:
    def bind_wrapper(res: Result[T, E]) -> Result[R, RE | E]:
        return Any

    # Function return type "Result[R@bind, RE@bind | E@bind_wrapper]" 
    # is incompatible with type "Result[R@bind, RE@bind | E@bind]"
    return wrap_function(bind_wrapper, func)
```

but its ok with variable 

```python 
value = wrap_function(bind_wrapper, func)
return value
```

In real code i use `Result` from https://github.com/rustedpy/result, but for example i do a minimal implementation.

**Comment by erictraut** [comment]
I presume that `R` and `RE`, which appear in the `bind` signature, are simple type variables. Those symbols are not defined in your example.

```python
R = TypeVar("R")
RE = TypeVar("RE")
```

Your code is running into an ambiguous case in the constraint solver. Any time you have a union between two type variables (like `RE | E`), you run the risk of unexpected behaviors. For example, consider that you assign a value of type `str | int | list` to a parameter of type `RE | E`. Should the constraint solver assume that `RE` captures `str` and `E` should capture `int | list`? Or should `RE` capture `str | int` and `E` should capture `list`? Or should `RE` capture `str | int | list` and `E` should go unsolved? A constraint solver needs to (somewhat arbitrarily) make these decisions. Pyright's constraint solver contains a number of heuristics that try to determine the "best" approach, but as with all heuristics, they can fail. That's what's happening here.

I'll look at this case a bit more closely to see if I can tweak the heuristics to get this case to work without breaking other cases. I just want to set expectations that this may be closed as "won't fix".

**Comment by minmax** [comment]
>I presume that R and RE, which appear in the bind signature, are simple type variables.
yes, just a TypeVar()

My understanding of how a type checker works is at a very low level, thank you for such detailed explanations. It's very helpful.

**Comment by erictraut** [comment]
My initial diagnosis was incorrect. I've confirmed this is a real bug, not an issue with a heuristic in the constraint solver.

The problem is in some logic that handles the case where a function-scoped TypeVar appears only within a callable in the return annotation. This is the case for `E` in the `bind` function. When pyright detects this case, it "rescopes" the type variable from the function to the return callable. This supports cases like the following:

```python
# We want `T` to be scoped to the return callable rather than
# function `func`. If it's scoped to `func`, then `func()` will be
# evaluated as `(Any) -> Any` rather than `(T) -> T` as the author
# most likely intended.
def func() -> Callable[[T], T]:
    def inner(x):
        return x
    return inner
```

In your example, `E` is used both in the outer function (`bind`) and the inner function (`bind_wrapper`), so it needs to remain scoped to `bind`. I'll need to think a bit more about how best to fix this.

**Comment by erictraut** [comment]
This is included in pyright 1.1.332, which I just published. It will also be included in a future release of pylance.
